[IV-22-23] Objetivo 5. Creación de un contenedor para pruebas

delightfulagony opened the pull request:
* [x] ¿Se han establecido criterios previos de elección de la imagen base, así
como los criterios de búsqueda que han conducido a las imágenes que se están
considerando?
* [x] ¿Se ha tenido en cuenta que la imagen es para testear, y que por tanto
tiene una función específica, tal como testear cómo va nuestra aplicación en la
última versión publicada del lenguage?
* [x] ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?
* [x] ¿Se han tenido en cuenta criterios reales para la elección, y no criterios
irrelevantes como cuanto tiempo se tarda en construirla? (Se va a construir una
vez, se va a usar cientos o miles de veces)
* [x] ¿Se han seguido las buenas prácticas en las órdenes que construyen
   la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?
* [x] ¿La imagen incluye sólo y exclusivamente la infraestructura
   necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
   "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
   comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
   nuestro proyecto?
* [ ] ¿Se han documentado y enlazado los commits a las imágenes que se
   han testeado?
* [x] ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los
   tests que se van a lanzar?
* [x] ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
en vez del directorio de trabajo?
* [x] ¿Se han establecido claramente los criterios de búsqueda y de aceptación de
   la imagen base?
* [x] ¿Tienes claro cuales son las condiciones en las que tiene que actualizarse
la imagen en DockerHub y has configurado el workflow de acuerdo con ello? En
particular, ¿sabes qué condiciones tienen que darse para que se actualice y qué
condiciones *no* deben darse?
* [x] Si se usan GitHub actions para su publicación, ¿se han considerado
cuidadosamene los criterios (ramas, paths) para que se lancen?
* [x] ¿Se ha comprobado que el ENTRYPOINT del contenedor no escriba nada en el
directorio que monta, ya que no tiene derecho de escritura?
* [x] ¿Se ha abierto un shell en el contenedor una vez construida la imagen y
comprobado que los ficheros que se necesitan están todos incluidos, y que no hay
ficheros que no se necesiten?

delightfulagony committed:
Fix #46: Construye imagen para CI

Se ha seleccionado la imagen base de rust:alpine que cumple con la
mayoría de requisitos preestablecidos:
- Peso de la imagen:
La imagen se descarga cada vez que se ejecutan los tests, por lo que es
importante que no sea muy pesada, de lo contrario los tiempos de test serán
muy largos lo que afecta negativamente al desarrollo.
- Velocidad (Muy ligada al peso):
Rust es un lenguaje notoriamente lento en compilación, por lo que al igual que
en el punto anterior interesa que la imagen sea lo más rápida posible para
facilitar el desarrollo.
- `stable` vs `edge`:
Se priorizará una imagen lo más actualizada posible, ya que no pasa nada si
falla interesará que puedan darse todos los problemas lo antes posible `fail
fast, fail early`. Al no estar en producción la estabilidad no es una
prioridad, más bien al contrario.
- Minimizar la superficie de ataque:
Dado que no es una imagen que vaya a usarse en producción, el objetivo es que
disponga de todo lo necesario para ejecutar los tests, no que su software sea
mínimo.
- Estable en el tiempo:
Esta es una imagen que se lanzará, ejecutará los tests y luego se eliminará,
por tanto no es necesario que pueda ejecutarse continuamente durante un periodo
de tiempo largo sin fallos.

No cumple:
- Disponibilidad de las herramientas necesarias:
Esto casi huelga decirlo, la imagen debe contener todas las herramientas
necesarias para contruir y ejecutar los tests de la aplicación.

Pero no se ha encontrado ninguna imagen que lo cumpla por lo que se
escoge la mejor opción encontrada.

delightfulagony committed:
Merge branch 'master' into objetivo-5

delightfulagony committed:
Revert "Merge branch 'master' into objetivo-5"

This reverts commit 40854d4ccdb15fb14056789cf1846ad894a967f9, reversing
changes made to b3f1163c3260dc9acb950a31ce80ccb8ab42ceef.

delightfulagony committed:
Actualiza README.md para reflejar el estado actual

delightfulagony committed:
Se cambia de nextest a cargo test

El motivo de este cambio es que nextest no respeta la variable de
entorno CARGO_TARGET_DIR al ejecutar sus tests y trata de escribir
igualmente en /target, imposibilitando construir correctamente una
imagen para ejecución de los tests.

delightfulagony committed:
Dockerfile más descriptivo e imagen más ligera

Se eliminan archivos innecesarios tras la construcción de la imagen

delightfulagony committed:
Fix #47: Añade una GA que actualiza oishi:testing

delightfulagony committed:
Fix #48: oishi:testing solo se actualiza en master

Añade una condición a la GA que automatiza la actualización de
oishi:testing para ejecutarse sólo cuando se realicen cambios a
Dockerfile en master.

delightfulagony committed:
Cambia :testing por :latest

La GA del repo de IV-22-23 espera que la etiqueta sea latest, así que se
cambia dicha etiqueta para que sea acorde a ésto.

delightfulagony committed:
Fix #50 por medio de un enlace simbólico

Se crea un enlace simbólico a /src para ejecutar los tests desde el
directorio /app que contiene un lockfile generado con la imagen.

delightfulagony commented:
@JJ Listo para revisión! :sweat_smile:

JJ commented:
Pero ¿lo has enviado?

delightfulagony commented:
¿En IV-22-23? Sí https://github.com/JJ/IV-22-23/pull/250

JJ commented on .github/workflows/test-image-push.yml:
```diff
@@ -0,0 +1,29 @@
+    branches:
+      - master
+    paths:
+      - "Dockerfile"
```
Habrá alguna otra cosa que cambie para crear la imagen, ¿no?

delightfulagony replied:
También se debería actualizar cuando cambien las dependencias! Claro :sweat_smile:

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+RUN mkdir -p /app/test
+
+# Install bash and curl for follow-up installation of nextest and justfile
+RUN apk add bash curl
```
`alpine` incluye `wget`. A veces peta, pero para bajar cosas no debería de ir mal.

JJ replied:
Tampoco necesitas bash; sh puede hacer la mayoría de las cosas.

delightfulagony replied:
Efectivamente no lo necesitaba, hasta el punto de que no lo estaba usando :disappointed: :laughing:

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+# Update all dependencies
+COPY --chown=tester Cargo.toml /app/
+COPY --chown=tester src/ /app/test/src/
```
Este directorio de va a montar; lo que hagas no le va a afectar. No puedes copiar los fuentes tampoco. Los fuentes se tienen que montar desde fuera.

delightfulagony replied:
Montaba src/ porque Cargo necesita un main.rs o lib.rs para actualizar (cosa que no entiendo muy bien pero bueno supongo que es la idiosincrasia de Rust, lo he solventado creando un main.rs vacío y borrándolo tras actualizar

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# Base image
+FROM rust:alpine
+
+# Intialize app directory for testing
```
No uses comentarios inline

delightfulagony replied:
Creo que ya está solventado

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+RUN curl --proto '=https' --tlsv1.2 -sSf https://just.systems/install.sh | bash -s -- --to /bin
+
+# Create non-priviledged user and cache folders
+RUN adduser -DH tester && chown -R tester /app \
```
¿Seguro que la imagen base no tiene ningún usuario?

delightfulagony replied:
Solucionado! Ahora las acciones no las ejecuta nadie :ghost: ! (Las ejecuta el usuario `nobody` que viene con la imagen)

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+RUN adduser -DH tester && chown -R tester /app \
+	&& mkdir -p /tmp/cache \
+	&& chown -R tester /tmp/cache \
+	&& chmod -R a+rwx /tmp/cache
```
¿Por qué necesitas esto además de cambiarle el usuario?

delightfulagony replied:
No lo necesitaba :face_in_clouds:

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,35 @@
+COPY --chown=tester src/ /app/test/src/
+# This is to prevent cargo from complaining about the lockfile not being
+# present when running the tests
+RUN ln -s /app/test/src /app/src
```
OK, esto es sorprendente...

delightfulagony replied:
Cargo requiere que los fuentes se encuentren en la carpeta root del proyecto, y esa carpeta es la carpeta donde se encuentra Cargo.toml y Cargo.lock, he intentando encontrar la forma de que eso no ocurra, y con CARGO_MANIFEST_PATH se supone que se puede hacer pero cuando lo hago no encuentra el lockfile e intenta escribir en /app/test, esto es lo mejor que se me ha ocurrido para solucionarlo

JJ reviewed (changes requested):
Sé que rust es complicado, pero pase lo que pase no pueden estar los fuentes en la imagen. Los fuentes se montan, pueden o no ser diferentes de los que haya en la imagen y por tanto hacerlos inválidos.

delightfulagony committed:
Improves #46: Mejora del Dockerfile

Se elimina la creación de un usuario y se usa uno propio de la imagen
base, se elimina la descarga e instalación de curl, bash, nextest y
just por ser innecesarias para el contenedor.

Se eliminan comentarios innecesarios y se añade un comentario
explicativo acerca de cómo se hace que cargo funcione correctamente.

delightfulagony committed:
Improve #47: También es importante Cargo.toml

Ahora la imagen se actualiza también cuando se hacen cambios a
Cargo.toml y no solo al Dockerfile

delightfulagony commented:
@JJ Listo! Había un poco de código muerto en el Dockerfile que había olvidado limpiar

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+RUN mkdir -p /app/test/
+
+RUN mkdir -p /app/cache \
```
Es conveniente que los comandos estos que cambian el sistema de ficheros vayan juntos.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+WORKDIR /app
+
+USER nobody
```
Er. Una cosa es que pongas un usuario sin privilegios, y otra que pongas uno sin ningún privilegio. Es muy probable que la imagen base tenga tal usuario; si no, lo creas.

delightfulagony replied:
Sería adecuado usar 'guest'? Si no lo fuera entonces tendría que crear uno si no me equivoco

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+RUN mkdir -p /app/test/
+
+RUN mkdir -p /app/cache \
+	&& chmod -R a+rwx /app/
```
También es conveniente que uses un usuario con los criterios adecuados en vez de dar todos los privilegios en un directorio. Aunque sea un contenedor de tests, tantos privilegios son una tentación.

delightfulagony replied:
Okay, le doy la propiedad de /app a 'guest' que es el usuario que voy a usar, sería eso correcto?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+USER nobody
+
+# Update the project dependencies
```
Ningún comentario en ningún sitio. Con que lo pongas en el issue y los mensajes de commit, suficiente.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+# or lib.rs file for updating so we create an empty one and delete it after
+# the update
+COPY Cargo.toml /app/
+RUN mkdir -p /app/test/src \
```
Aunque vayas a borrarlo, este subdirectorio es mejor que no lo toques, porque se va a monter el de los fuentes. Si es temporal, usa un directorio temporal, pero no este.

JJ reviewed (changes requested):
Dale todavía una vueltecilla, por favor. Todavía hay algunas buenas prácticas que no has seguido.

delightfulagony committed:
Improve #46: Mejores prácticas

Se eliminan los comentarios, se pasa a usar el usuario 'guest' en vez de
'nobody' y se evita tocar el directorio '/app/test' donde se montarán
los fuentes, en lugar de éste se usa '/app/src' directamente.

delightfulagony commented:
@JJ Ya solventé lo que me dijiste, lo único que no tengo muy claro es si 'guest' es un usuario adecuado para la tarea o debería crear uno y listo

JJ commented:
> @JJ Ya solventé lo que me dijiste, lo único que no tengo muy claro es si 'guest' es un usuario adecuado para la tarea o debería crear uno y listo

Si el base tiene uno definido, ese me vale. Si no, tendrás que definir un usuario normal. No sé qué privilegios tendrá esa cuenta. La cuestión es que sea un usuario normal, con privilegios conocidos. Si no, crea uno con un nombre relacionado con el proyecto.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+	&& touch /app/src/main.rs \
+	&& cargo update \
+	&& rm -rf /app/src \
+	&& ln -s /app/test/src /app/src
```
Si lo estoy entendiendo bien, estás creando un directorio root "de mentira" que equivale al directorio `src` que se monta de forma externa. Como es un link simbólico, se poblará "automágicamente" cuando se monte, y como la mejor práctica es usar `src` para los fuentes, siempre van a estar ahí. Dejas el directorio de trabajo en `/app`
A ver, esto es una ñapa buena. 
El principal problema que veo es que puede que `Cargo.toml` no va a estar sincronizado en un momento si cambias los fuentes y este a la vez: en el mismo push se creará la imagen y se hará la prueba, pero con la imagen del `Cargo.toml` anterior.

delightfulagony replied:
En problema va más allá, Cargo **se niega** a actualizar sus dependencias o hacer nada si no tiene un archivo main.rs accesible desde su root, da igual que éste archivo esté vació o no, por eso hago el touch main.rs.

El siguiente problema es que, como en Cargo **exige** que los fuentes estén en /src, si coloco la carpeta root en /app entonces cargo no es capaz de encontrar los fuentes (puesto que la carpeta está en /app/test/src y no en /app/src/) pero si la carpeta root es /app/test entonces no encuentra el lockfile (porque esa carpeta no lo tiene) y trata de crearlo, abortando porque no tiene permisos de escritura.

delightfulagony commented:
> > @JJ Ya solventé lo que me dijiste, lo único que no tengo muy claro es si 'guest' es un usuario adecuado para la tarea o debería crear uno y listo
> 
> Si el base tiene uno definido, ese me vale. Si no, tendrás que definir un usuario normal. No sé qué privilegios tendrá esa cuenta. La cuestión es que sea un usuario normal, con privilegios conocidos. Si no, crea uno con un nombre relacionado con el proyecto.

'guest' es un usuario de invitados del grupo 'users' que viene con la imagen, no tiene home pero hasta donde sé debería tener privilegios normales, examinando el contenedor tiene permisos de ejecución, de red, no tiene permisos de %wheel (no está en el grupo ni la máquina tiene instalado sudo), puede crear y eliminar ficheros en las carpetas que le pertenecen y poco más, lo que viene siendo un usuario normal en linux diría yo

JJ reviewed (approved):
Te lo voy a dejar pasar, porque si mal no recuerdo, lo que hacían otros con Rust era copiar los fuentes dentro del contenedor y trabajar sobre ellos. El enlace simbólico es una solución más rápida, al no tener que copiar una cantidad de ficheros indeterminada. Tiene el problema de la dessincronicación de `cargo.toml` que te he indicado, lo que se podría solucionar desde la infraestructura, por ejemplo haciendo depender un workflow de otro. Sin embargo, eso es del objetivo 6 y la solución, sin ser perfecta, es aceptable (y más a estas alturas)

JJ commented:
Por favor, apúntate a la presentación y ven mañana, que ya sabes que no es suficiente con esta nota para llegar al 5.

JJ commented:
> > Si el base tiene uno definido, ese me vale. Si no, tendrás que definir un usuario normal. No sé qué privilegios tendrá esa cuenta. La cuestión es que sea un usuario normal, con privilegios conocidos. Si no, crea uno con un nombre relacionado con el proyecto.
> 
> 'guest' es un usuario de invitados del grupo 'users' que viene con la imagen, no tiene home pero hasta donde sé debería tener privilegios normales, examinando el contenedor tiene permisos de ejecución, de red, no tiene permisos de '%wheel' (no está en el grupo ni la máquina tiene instalado sudo), puede crear y eliminar ficheros en las carpetas que le pertenecen y poco más, lo que viene siendo un usuario normal en linux diría yo

Si, es posible. Sería de todas formas mejor que tuviera un usuario verdaderamente normal, con el nombre del proyecto. Si quieres hacerlo antes de mergear, mejor. Si no, no hay problema.

delightfulagony committed:
Improve #46: Cambia usuario por oishi

Crea un usuario llamado oishi para la imagen de testing en lugar de usar
guest
